### PR TITLE
spec: Implement ManifestList read functionality

### LINKED ISSUE #57: Implement ManifestList read functionality
First, support to read/parse the avro manifestList file

### PR DESCRIPTION
**Pull request by Aitozi**
close #57 

The test fixtures is generated by 

```
    public void testWriteManifestList() {
        List<ManifestFileMeta> metas = new ArrayList<>();

        BinaryRow row = new BinaryRow(2);
        BinaryRowWriter writer = new BinaryRowWriter(row);
        writer.writeInt(0,1);
        writer.writeString(1, BinaryString.fromString("1"));
        writer.complete();
        byte[] bytes1 = row.toBytes();

        BinaryArray array = BinaryArray.fromLongArray(new Long[] {1L, 2L});
        byte[] bytes2 = array.toBytes();

        metas.add(new ManifestFileMeta(
                "manifest-" + UUID.randomUUID(),
                10,
                10,
                10,
                new BinaryTableStats(row, row, array), 1));
        metas.add(new ManifestFileMeta(
                "manifest-" + UUID.randomUUID(),
                11,
                0,
                10,
                new BinaryTableStats(row, row, array), 2));

        ManifestList manifestList = createManifestList(tempDir.toString());
        String manifestListName = manifestList.write(metas);
}
```

### CODE REVIEW

File: crates/paimon/src/spec/manifest_list.rs [hunk 1]
```diff
@@ -16,15 +16,134 @@
...
+            .is_empty());
+
+        assert_eq!(
+            res.entries().get(1).unwrap().file_name(),
+            "manifest-88fe5442-3374-4385-a08f-022abd823a27"
+        );
+        assert_eq!(res.entries().get(1).unwrap().version(), 2);
+        assert_eq!(res.entries().get(1).unwrap().file_size(), 600);
+        assert_eq!(res.entries().get(1).unwrap().num_added_files(), 6);
+        assert_eq!(res.entries().get(1).unwrap().num_deleted_files(), 0);
```
Comment (XuQianJin-Stars) [inline crates/paimon/src/spec/manifest_list.rs:125]: Use a file with a manifest-list that contains `num_deleted_files`?

Comment (Aitozi) [reply crates/paimon/src/spec/manifest_list.rs:125]: Sure, will update.

Comment (Aitozi) [reply crates/paimon/src/spec/manifest_list.rs]: Updated

**Comment by Aitozi** [comment]
CC @Xuanwo , please help take a look when you are free, thanks

File: crates/paimon/src/error.rs [hunk 2]
```diff
@@ -52,6 +52,14 @@ pub enum Error {
         display("Paimon hitting invalid config: {}", message)
     )]
     ConfigInvalid { message: String },
+    #[snafu(
+        visibility(pub(crate)),
+        display("Paimon hitting unexpected avro error {}: {:?}", message, source)
+    )]
+    AvroFailed {
```
Comment (Xuanwo) [inline crates/paimon/src/error.rs:59]: I believe users don't need to be aware of `AvroFailed`. How about using `DataInvalid` or `DataUnexpected` until we determine which is more appropriate?

File: crates/paimon/src/spec/manifest_list.rs [hunk 3]
```diff
@@ -16,15 +16,133 @@
...
+        let path = workdir
+            .join("tests/fixtures/manifest/manifest-list-5c7399a0-46ae-4a5e-9c13-3ab07212cdb6-0");
+        let v = std::fs::read(path.to_str().unwrap()).unwrap();
+        let res = ManifestList::from_avro_bytes(&v).unwrap();
+        assert_eq!(res.entries().len(), 2);
+        assert_eq!(
+            res.entries().first().unwrap().file_name(),
+            "manifest-7ff677cc-46c0-497d-8feb-1c6785707a4b"
+        );
+        assert_eq!(res.entries().first().unwrap().version(), 2);
```
Comment (Xuanwo) [inline crates/paimon/src/spec/manifest_list.rs:89]: How about `assert_eq!(res, ManifestList { ... })` directly? I think we can derive `Eq` for ManifestList.

File: crates/paimon/src/error.rs (hunk 2 above)
Comment (Aitozi) [reply crates/paimon/src/error.rs:59]: It's better, updated.

File: crates/paimon/src/spec/manifest_list.rs (hunk 3 above)
Comment (Aitozi) [reply crates/paimon/src/spec/manifest_list.rs]: Updated

**Review by Xuanwo — APPROVED** [review]
Thanks!

**Comment by Xuanwo** [comment]
Hi, @JingsongLi @SteNicholas, this PR is happy to go.

**Review by XuQianJin-Stars — APPROVED** [review]
+1

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**